tidy3d/components/dispersion_fitter.py

```diff
@@ -0,0 +1,936 @@
+"""Implementation of pole-residue dispersion fitter."""
+
+from __future__ import annotations
+
+from typing import Optional, Tuple
+
+import numpy as np
+import scipy
+from pydantic.v1 import Field, NonNegativeFloat, PositiveFloat, PositiveInt, validator
+from rich.progress import Progress
+
+from ..constants import fp_eps
+from ..exceptions import ValidationError
+from ..log import get_logging_console, log
+from .base import Tidy3dBaseModel, cached_property, skip_if_fields_missing
+from .types import ArrayComplex1D, ArrayComplex2D, ArrayFloat1D, ArrayFloat2D
+
+# numerical tolerance for pole relocation for fast fitter
+TOL = 1e-8
+# numerical cutoff for passivity testing
+CUTOFF = np.finfo(np.float32).eps
+# parameters for passivity optimization
+PASSIVITY_NUM_ITERS_DEFAULT = 50
+SLSQP_CONSTRAINT_SCALE_DEFAULT = 1e35
+# min value of the rms for default weights calculated based on Re[eps], Im[eps]
+RMS_MIN = 0.1
+
+DEFAULT_MAX_POLES = 5
+DEFAULT_NUM_ITERS = 20
+DEFAULT_TOLERANCE_RMS = 1e-5
+
+# this avoids divide by zero errors with lossless poles
+SCALE_FACTOR = 1.01
+
+# when poles are close to omega, can cause invalid response function, and we reject model
+OMEGA_POLE_CLOSE_ATOL = 1e-10
+
+# Range for checking upper bound of Im[eps], in addition to extrema method.
+# The range is in unit of eV and it's in log scale.
+LOSS_CHECK_MIN = -10
+LOSS_CHECK_MAX = 4
+LOSS_CHECK_NUM = 1000
+
+
+def imag_ep_extrema(poles: ArrayComplex1D, residues: ArrayComplex1D) -> ArrayFloat1D:
```

Comment:
what is `ep`? might want to consider giving this function a more intuitive name.